tensorflow/python/ops/dynamic_embedding_ops.py

```diff
@@ -0,0 +1,873 @@
+    self.name = name
+    self.shared_name = shared_name or "shared_name.{}".format(name)
+
+    self.initializer = None
+
+    self.trainable = trainable
+    self.checkpoint = checkpoint
+
+    self._tables = []
+    self.size_ops = []
+    self.shard_num = len(self.devices)
+
+    if key_dtype not in [dtypes.int32, dtypes.int64]:
+      raise TypeError("key_dtype should be int32, int64.")
+    _initializer = initializer
+    if _initializer is None:
+      _initializer = init_ops.zeros_initializer(dtype=self.value_dtype)
+    static_default_value = self._convert_anything_to_init(_initializer, dim)
+
+    value_dtype_list = [
+        dtypes.int32, dtypes.int64, dtypes.bool, dtypes.float32, dtypes.float64
+    ]
+    if value_dtype not in value_dtype_list:
+      raise TypeError("value_dtype should be ", value_dtype_list)
+    scope_name = self.name.split("/")[-1]
+    with ops.name_scope(scope_name, "DynamicEmbedding_Variable"):
+      with ops.colocate_with(None, ignore_existing=True):
+        for idx in range(len(self.devices)):
+          with ops.device(self.devices[idx]):
+            mht = None
+            mht = lookup_ops.MutableHashTable(
+                key_dtype=self.key_dtype,
+                value_dtype=self.value_dtype,
+                default_value=static_default_value,
+                name=self._make_name(idx),
+                checkpoint=self.checkpoint)
+
+            self._tables.append(mht)
+    super(Variable, self).__init__()
+
+  @property
+  def tables(self):
+    return self._tables
+
+  def _convert_anything_to_init(self, raw_init, dim):
+    init = raw_init
+    while callable(init):
+      if isinstance(init, (init_ops.Initializer, init_ops_v2.Initializer)):
+        self.initializer = init
+        init = init(shape=[1])
+      else:
+        init = init()
+    init = math_ops.cast(array_ops.fill([dim],
+                                        array_ops.reshape(init, [-1])[0]),
+                         dtype=self.value_dtype)
+    return init
+
+  def _create_resource(self):
+    raise NotImplementedError
+
+  def _make_name(self, table_idx):
+    return "{}_mht_{}of{}".format(self.name.replace("/", "_"), table_idx + 1,
+                                  self.shard_num)
+
+  def upsert(self, keys, values, name=None):
+    """Insert or Update `keys` with `values`.
+
+    If key exists already, value will be updated.
+
+    Args:
+      keys: Keys to insert. Can be a tensor of any shape. Must match the table's
+        key type.
+      values: Values to be associated with keys. Must be a tensor of the same
+        shape as `keys` and match the table's value type.
+      name: A name for the operation (optional).
+
+    Returns:
+      The created Operation.
+
+    Raises:
+      TypeError: when `keys` or `values` doesn't match the table data
+        types.
+    """
+
+    partition_index = self.partition_fn(keys, self.shard_num)
+    keys_partitions, _ = _partition(keys, partition_index, self.shard_num)
```

Comment:
Suppose we have keys in range [0, 10000] and shard_num = 8, then partition_0 will got keys [0, 8, 16, 24, ...., 10000], and MutableHashTableOfTensors will store these keys in a std::unordered_map container.
However, unordered_map uses collision chaining to resolve hash collisions and slot_index = hash(key) % bucket_count, suppose bucket_count was huge like 2^20, thus keys' slot_index will be [0, 8, 16, 24, ...., 10000], that means we only use 1/shard_num slots of bucket_count in unordered_map, this may leader huge hash collision, and finally leader to poor performance.
Suggestion:
1. provide a int/int64 hash function like murmur hash
2. one-to-one mapping keys to another format like new_keys = keys / shard_num

Reply:
Hi @rangjiaheng ,thanks for your comments, that's really a problem  and I will consider your suggestion, but before that maybe you can customize a `partitioner` for `Variable` to avoid this problem.

Reply:
@rangjiaheng  Any update on this PR? Please. Thanks!